### PR TITLE
[Event] Add EVENT_CLI_SUCCESSFUL_EXECUTE

### PR DESCRIPTION
**Pull request by jiasli**
## Context

In the current implementation, the command handler is invoked at `L215`:

https://github.com/microsoft/knack/blob/0d0308bc567f2d25654bb39201c02ab262d7ce0a/knack/cli.py#L215

The JSON is printed at `L221`:

https://github.com/microsoft/knack/blob/8b3757ddff6b63bf4789c8736a506e1754f1f92b/knack/cli.py#L221

Therefore, anything printed by the command handler will be shown **before** the JSON output.

## Change

Add a new event `EVENT_CLI_SUCCESSFUL_EXECUTE`. It is raised after a command is successfully executed, allowing a callback `func` to be registered and called after JSON is printed. 

This enables post-output hint to be displayed **after** the JSON output.

### CODE REVIEW

**Review by evelyn-ys — APPROVED** [review]

File: knack/cli.py [hunk 1]
```diff
@@ -219,6 +219,7 @@ def invoke(self, args, initial_invocation_data=None, out_file=None):
                 if cmd_result and cmd_result.result is not None:
                     formatter = self.output.get_formatter(output_type)
                     self.output.out(cmd_result, formatter=formatter, out_file=out_file)
+            self.raise_event(EVENT_CLI_SUCCESSFUL_EXECUTE)
```
Comment (Juliehzl) [inline knack/cli.py:222]: two concerns:
1. could the new event bring benefit to CLI if exit code = 0 could represent successful execution?
2.  should SUCCESSFUL_EXECUTE be raised after cmd_result is returned? It seems that in current code output format failure would block the event.

Comment (jiasli) [reply knack/cli.py:222]: > could the new event bring benefit to CLI if exit code = 0 could represent successful execution?

I couldn't really get your question. Could you elaborate?

> should SUCCESSFUL_EXECUTE be raised after cmd_result is returned?

No. If `self.output.out` failed, this is considered a failure and `SUCCESSFUL_EXECUTE` shouldn't be raised by design.

File: knack/cli.py [hunk 2]
```diff
@@ -219,6 +219,7 @@ def invoke(self, args, initial_invocation_data=None, out_file=None):
                 if cmd_result and cmd_result.result is not None:
                     formatter = self.output.get_formatter(output_type)
                     self.output.out(cmd_result, formatter=formatter, out_file=out_file)
+                self.raise_event(EVENT_CLI_SUCCESSFUL_EXECUTE, result=cmd_result.result)
```
Comment (fengzhou-msft) [inline knack/cli.py:222]: Should we raise the event only when exit_code is 0?

Comment (jiasli) [reply knack/cli.py:222]: Yes. On the other hand, `EVENT_CLI_POST_EXECUTE` is raised unconditionally. 

https://github.com/microsoft/knack/blob/61057a3a92756f01701c9428959a2bc1185559bf/knack/cli.py#L233-L234

Comment (fengzhou-msft) [reply knack/cli.py:222]: Then should you check the `exit_code` from https://github.com/microsoft/knack/blob/61057a3a92756f01701c9428959a2bc1185559bf/knack/cli.py#L217

Comment (jiasli) [reply knack/cli.py:222]: Not really necessary. 

https://github.com/microsoft/knack/blob/61057a3a92756f01701c9428959a2bc1185559bf/knack/cli.py#L215

will raise an exception first if `exit_code` is not `0`. 

Also,

https://github.com/microsoft/knack/blob/61057a3a92756f01701c9428959a2bc1185559bf/knack/cli.py#L221

already assumes the command is successfully executed and `exit_code` is `0`.

Comment (fengzhou-msft) [reply knack/cli.py:222]: got it.

**Review by fengzhou-msft — APPROVED** [review]

**Review by Juliehzl — APPROVED** [review]

**Pull request merged**